### ISSUE TITLE
NotImplementedError when training FarSeg model from torchgeo.models

### ISSUE THREAD

**Issue by KennSmithDS** [microsoft/torchgeo#301]
**Description of Issue**
When attempting to use one of the models built into torchgeo, `from torchgeo.models import FarSeg` I receive a `NotImplementedError` when running the cell in Jupyter Notebook that performs the fitting of the model using a PyTorch Lightning `Trainer`.
I have uploaded the source code for the CloudFarSegModel class and the CloudDataset class to this repository [here](https://github.com/KennSmithDS/drivendata_cloud_competition). The competition dataset is nearly 30 gb so it would have to be downloaded from the AzureBlobStorage source.

[FarSeg_NotImplementedError.txt](https://github.com/microsoft/torchgeo/files/7738090/FarSeg_NotImplementedError.txt)
![NotImplementedError](https://user-images.githubusercontent.com/30271891/146615857-341726f8-6039-4ac2-8ada-55be44ab7f88.png)

**Credit to Katie Wetstone for source code from the blog/notebook to train a benchmark Unet model on the dataset:**
https://www.drivendata.co/blog/cloud-cover-benchmark/

**Comment by adamjstewart** [comment]
The `model` argument to `pl.Trainer` should be a `pl.LightningModule`, not an `nn.Module`. Using tools like `mypy` can help find these bugs, although I'm not sure how they work with Jupyter Notebook. If you look at our [PyTorch Lightning trainers tutorial](https://torchgeo.readthedocs.io/en/stable/tutorials/trainers.html), you'll see we instantiate a `RegressionTask` and pass that to the `trainer.fit` method.

**Comment by isaaccorley** [comment]
Your error message indicates that your model has no `forward` method. Looking at your `CloudFarSegModel` implementation [here](https://github.com/KennSmithDS/drivendata_cloud_competition/blob/main/farseg_src/cloud_model.py) shows that you don't have one.

**Comment by KennSmithDS** [comment]
Hi Adam, 

Appreciate the quick response!

Yes, I had looked through that tutorial and saw the `RegressionTask`, and thought to implement the `SemanticSegmentationTask` but didn't see where I would instantiate a model like `FarSeg`. I see what you mean now from looking at torchgeo source code that `SemanticSegmentationTask` is a `pl.LightningModule`, where FarSeg is an `nn.Module`. 

Is there an example on how to configure a SemanticSegmentationTask` or would something like this work? Conversely, example of how I'd use the FarSeg model for semantic segmentation?

`task = SemanticSegmentationTask(
      segmentation_model="deeplabv3+",
      encoder_name="resnetXX",
      encoder_weights="imagenet",
      in_channels=4,
      num_class=2
      loss= [is this a string or a loss object?]
)
`

Cheers

**Comment by adamjstewart** [comment]
@KennSmithDS the trainer stuff isn't as well-documented or extensible as I would like. If you look at the source code for `SemanticSegmentationTask`, look in the `config_task` function to see what you need to do. I think you can do something like:
```python
class MySemanticSegmentationTask(SemanticSegmentationTask):
    def config_task(self) -> None:
        self.model = FarSeg(...)  # or a different model
        self.loss = CrossEntropyLoss(...)  # or a different loss object
```

**Comment by isaaccorley** [comment]
@adamjstewart's solution above should work for you. To answer your question about the loss, it's a string that, as of now, can be one of ["ce", "jaccard", "focal"] which corresponds to `torch.nn.CrossEntropyLoss`, `segmentation_models_pytorch.losses.JaccardLoss`, and `segmentation_models_pytorch.losses.FocalLoss`, respectively.

**Comment by KennSmithDS** [comment]
@adamjstewart and @isaaccorley thank you both!

If I go the route of using the existing SemanticSegmentationTask, do I need to create a custom data module for this cloud cover dataset then to pass into the `datamodule=` parameter for the Trainer?

Otherwise extending the SemanticSegmentationTask for a FarSeg (or other) model makes sense too.

**Comment by calebrob6** [comment]
> do I need to create a custom data module for this cloud cover dataset then to pass into the datamodule= parameter for the Trainer?

Yes, exactly!

**Comment by isaaccorley** [comment]
@KennSmithDS feel free to make a separate issue if you run into any problems with training on your custom datamodule using one of ours as template so we can document this for other users. 

Also feel free to make a PR for the the Cloud Cover challenge dataset and datamodule. It would be a good first contribution and it would be great to have it included in torchgeo.
